DM-33764: First draft.

TallJimbo opened the pull request.

TallJimbo committed:
Add introduction text and outline as empty sections.

timj commented on index.rst:
```diff
@@ -1,62 +1,147 @@
-  ^^^^^^^^^^^^^^
+DMTN-181 :cite:`DMTN-181` describes a Campaign Management system that provides tooling for organizing and tracking the multiple workflows that comprise a pipeline processing campaign.
+It also touches on what we will here call (for lack of a better term, so far) *Campaign Definition*: the tools and processes for determining the datasets, software, and configuration that will go into a campaign.
+The datasets are our primary concern here, and for these :cite:`DMTN-181` focuses on the concept of exposure *exclusion lists* represented as *data ID sets*.
```
There's also how this fits in with summit users creating tags and quality flags for datasets (see also DMTN-204) and where that information ends up in USDF.

TallJimbo replied:
There's now a later section where I touch on this (and reference DMTN-204), but I'm probably not the right person to write much more about it.  I'm happy to take text and certainly corrections on what I've said about dataset annotations, as it's something that I know is related, but I don't know much about the detailed use cases and don't have strong opinions about how it's implemented as a result.

timj commented on index.rst:
```diff
@@ -1,62 +1,147 @@
+.. note::
 
-     Caption text.
+   The Consolidated Database probably has a role to play here; perhaps it is the database that should support Campaign Definition, and perhaps it is even a database that we should consider merging with the `Registry`.
```
Consolidated Database means "everything that we want to store in a database" doesn't it? Even for the Exposure table if we integrated it into registry it would mean at least one special table per instrument where the schema is essentially arbitrary since what we can store in it is unlimited and can come from all the fits headers and EFD. At least the ObsCore table is a well-defined schema and is directly related to the Exposure records.

TallJimbo replied:
> Consolidated Database means "everything that we want to store in a database" doesn't it?

Exactly why I didn't want to try to include it here - as long as that continues to be what it means, it'll continue to be vaporware.  Once it tries to do less we'll have a sense of what it can do.

timj commented on index.rst:
```diff
@@ -1,62 +1,147 @@
+The processes and tools for creating and maintaining exclusion lists are considered out of scope for campaign management by :cite:`DMTN-181`, but it correctly notes that these must be assembled from many disparate inputs that cannot be assumed to be present in any single database (let along the `Registry` database), so generating exclusion lists from these inputs implicitly as part of the queries ...
+Defining those processes and tools fully is out of scope for this technote, too, but we will try to anticipate their demands on the middleware.
+That includes support for using the data repository as a place to store and query metrics, dataset annotations, and other criteria that play a role in building exclusion lists, as well as making the `Registry` database the source of truth for curated data ID sets (such as the exclusion lists themselves).
+It should be emphasized that this support has not been requested by those involved in pushing for and designing the Campaign Management system; on the contrary, they regard relying on the `Registry` database as over-centralization and poor separation of concerns.
```
who are "those involved in pushing for and designing the Campaign Management system"? RTN-023 talks about Campaign Tooling but this document has to be clear it's talking about something completely different. Campaign Management System being a different thing to Campaign Tooling  for Generating, Monitoring, and Tracking data processing Campaigns (or defining their relationship) is key to avoiding continued confusion.

TallJimbo replied:
I've definitely heard this sentiment expressed most forcefully from @RobertLuptonTheGood , but it is also a general sense I've gotten in discussions about Campaign Management.  It may be that I'm subconsciously conflating [my own interpretation of] Robert's opinions about architecture with a distinct set of concerns from others about middleware taking on too much scope, especially in construction.

timj commented on index.rst:
```diff
@@ -1,62 +1,147 @@
-  _static/ directory. The reST syntax for adding the image is
+The processes and tools for creating and maintaining exclusion lists are considered out of scope for campaign management by :cite:`DMTN-181`, but it correctly notes that these must be assembled from many disparate inputs that cannot be assumed to be present in any single database (let along the `Registry` database), so generating exclusion lists from these inputs implicitly as part of the queries ...
+Defining those processes and tools fully is out of scope for this technote, too, but we will try to anticipate their demands on the middleware.
+That includes support for using the data repository as a place to store and query metrics, dataset annotations, and other criteria that play a role in building exclusion lists, as well as making the `Registry` database the source of truth for curated data ID sets (such as the exclusion lists themselves).
```
Cite metrics and dataset annotations DMTNs.

TallJimbo replied:
I've got specific subsections for these later where I was planning to make those references, but it doesn't hurt to cite them up here, too.

timj commented on index.rst:
```diff
@@ -1,62 +1,147 @@
+Defining those processes and tools fully is out of scope for this technote, too, but we will try to anticipate their demands on the middleware.
+That includes support for using the data repository as a place to store and query metrics, dataset annotations, and other criteria that play a role in building exclusion lists, as well as making the `Registry` database the source of truth for curated data ID sets (such as the exclusion lists themselves).
+It should be emphasized that this support has not been requested by those involved in pushing for and designing the Campaign Management system; on the contrary, they regard relying on the `Registry` database as over-centralization and poor separation of concerns.
+These are valid worries - it is fair to say that the butler tries to do too much already, schema stability and migration are not trivial problems, and the middleware codebase currently has few contributors - but if the alternative is standing up, populating, and maintaining another SQL database with content similar to or substantially overlapping the `Registry`'s, then tighter coupling between mi ...
```
Part of the reason I pushed Jira in RTN-023 was to stop us immediately jumping into needing an entirely new Postgres database with all the associated sqlalchemy and schema migration problems.

timj reviewed (commented):
My main request is we clarify the relationship between this document and RTN-023. The nomenclature overload is something we need to sort out. RTN-023 mainly concerns itself with tooling once the batch jobs have been defined.

TallJimbo committed:
Add draft section saying what's possible now.

TallJimbo committed:
Fill out section describing possible new middleware features.

I've dropped two that were present in the outline:

 - Thinking about it further, providing a public SQL interface to the
 registry still seems hard, and since we have no way to query via the
 registry and extension tables added manually to the database, I don't
 really see a problem this solves.

 - An in-memory query engine for QG generation is still an interesting
 idea to me, but I don't think it's relevant for Campaign Management.

TallJimbo committed:
Add section on other drivers for these features.

TallJimbo committed:
Update bibliographies.

TallJimbo committed:
Fix links and citations to other DMTNs.

TallJimbo committed:
Full-read proofing adjustments and fill in recommendations.

ctslater commented on index.rst:
```diff
@@ -1,62 +1,427 @@
+
+Based on current middleware capabilities, the perceived difficulty of extending those capabilities in the ways described above, and our guesses at how difficult it would be to stand up and maintain non-middleware (or less-middleware-based) solutions to various problems, we make the following recommendations for Campaign Management/Definition:
+
+#. We should mandate a one-to-one relationship between ``RUN`` collections and workflows.
```
This sounds really great, I think we're in-practice converging in that direction but it's nice to make it a clear requirement. 

What I'm not sure about is how that per-workflow data is accessed once the run collection is chained together with other neighboring or downstream workflows? It seems like users or the campaign system will get quasi-random results, depending on the not-meaningful order of the chaining. Not a huge problem if one can walk the chain back to the runs, but it seems like a case where datasets shadowing "older" versions with the same dimensions isn't the ideal behavior.

TallJimbo replied:
That's a good point, and I don't have a solid answer right now.  Certainly the right thing to do to get any per-RUN datasets is to use the RUN to query for them, and use the CHAINED collection only to find the RUNs, which is possible (or in the future, we could use quantum provenance to get them). But I agree it would be nice to make it hard to accidentally get only some arbitrary, semi-latest one by querying the CHAINED collection directly.  I can think of a few possible solutions that wouldn't be hard to implement, but would also involve some low-level special-casing, and I'd prefer to come up with a more general mechanism if I can.

ctslater commented on index.rst:
```diff
@@ -1,62 +1,427 @@
+   This allows per-workflow provenance and metadata to be saved as butler datasets without new dimensions, and it automatically associates that per-workflow information with the datasets produced by the workflow.
+   If this leads too problems with query performance due to too many ``RUN`` collections in certain ``CHAINED`` collections, we have a few avenues of optimization we can explore to address it.
+
+#. We should use ``TAGGED`` collections referenced through per-instrument ``CHAINED`` collection pointers to maintain the official, current-best ``raw`` exclusion list (as an inclusion list).
```
At first glance, the size of such an inclusion list makes me scared about performance? Do you think this is not a worry, or a risk we can manage?

TallJimbo replied:
My mental model is that storing an inclusion list vs. an exclusion is a performance vs. storage tradeoff, based on the assumption (which may be incorrect) is that the more complicated queries an exclusion list would require would be slower.  But I also think it isn't hard to implement a "set difference" type of collection to allow us to store an exclusion list directly in the future, if we have to.

ctslater commented on index.rst:
```diff
@@ -1,62 +1,427 @@
+   Whenever Campaign Definition processes consume observational metadata from a non-butler data source, however, we should at least consider whether it or some summary of it should be included in the data repository, either as a dataset or a dimension column, especially when the production resembles a major operations one, like data release or alert production (as opposed to ad-hoc hardware commi ...
+   Campaign Definition queries will often be good predictor of things science users or pipelines developers will want to do, and we don't want that to involve too many heterogeneous sources of data.
+
+#. Similarly, when standing up new systems to track and store observational metadata or Campaign Management state (or deciding whether to stand up a new one or piggyback on something else), we should prefer *not* to use a butler data repository when the organizational structure is extremely simple (e.g. strictly per exposure or per workflow) or organized along dimensions not in the butler data mo ...
```
I'm having a hard time understanding the distinction between what recommendations 1 and 3 propose should be in the butler versus what this is saying should not be. Per-workflow information seems to be prohibited by this and allowed by 1? Maybe I should read the second half of this recommendation as taking precedence over the first half?

TallJimbo replied:
I think there's a genuine tension here for per-workflow information, and I hadn't fully appreciated it when writing these (I think I may have written the bullets on different days :slightly_smiling_face: ):

 - Using the butler to store per-workflow information doesn't really make it easier to organize that information on its own, because the organizational structure is already simple, and of course using the butler puts some constraints on other aspects of the implementation.
 - Using the butler to store per-workflow information (if workflows are 1-1 with RUN collections) does make it a lot easier to relate that information to datasets, quanta, and especially per-RUN information handled by the middleware.

Resolving that tension comes down to how much being tightly coupled to the butler is a problem, how much you care about linkage to existing butler information, and how you want to store information that isn't *naturally* well-organized by butler (like per-campaign and per-job - I think it's *possible* to organize those well within the butler, but it's not something we'd provide out of the box).

TallJimbo committed:
Address a few early review comments.
